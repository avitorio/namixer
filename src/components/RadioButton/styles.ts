import styled, { css } from 'styled-components'
import media from 'styled-media-query'
import { lighten } from 'polished'

export const Wrapper = styled.div`
  ${() => css`
    display: flex;
    align-items: center;
    position: relative;

    ${media.greaterThan('small')`
      &:hover .tooltip {
        display: block;
      }
    `}
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    opacity: 0;
    position: fixed;
    width: 0px;

    &:checked + label {
      border: 2px solid ${theme.accentColors.green};
      background: ${theme.colors.secondary};
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    align-items: center;
    background-color: ${theme.colors.neutral.neutral100};
    border: 2px solid #ffffff00;
    border-radius: ${theme.border.radius.small};
    color: ${theme.colors.neutral.neutral900};
    cursor: pointer;
    display: inline-flex;
    justify-content: center;
    font-family: ${theme.font.family};
    font-weight: ${theme.font.bold};
    padding: 1rem;
    text-decoration: none;
    position: relative;
    white-space: pre;

    &:hover {
      z-index: 4;
      background-color: ${lighten(0.05, theme.colors.secondary)};

      &.tooltip:after {
        content: '';
        position: absolute;
        bottom: -23px;
        left: 50%;
        margin-left: -16px;
        border-width: 15px;
        border-style: solid;
        border-color: transparent transparent #fff;
        z-index: 2;
      }
    }

    ${media.greaterThan('medium')`
      padding: ${theme.spacings.xsmall};
    `}
  `}
`

export const Tooltip = styled.div`
  ${({ theme }) => css`
    display: none;
    position: absolute;
    left: 0;
    width: 500px;
    max-width: 90vw;
    z-index: 3;
    top: 100%;
    line-height: 1.7;
    color: ${theme.colors.neutral.neutral900};

    > div {
      margin-top: 20px;
      display: inline-block;
      background: white;
      padding: 10px 15px;
      box-shadow: ${theme.shadow.fluffy.elevation5};
      border-radius: 4px;
    }
  `}
`
