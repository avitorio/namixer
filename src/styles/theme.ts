export default {
  grid: {
    container: '137rem',
    gutter: '3.2rem'
  },
  border: {
    radius: {
      small: '0.4rem',
      big: '0.8rem'
    }
  },
  font: {
    family:
      "Source Sans Pro, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    light: 300,
    normal: 400,
    bold: 600,
    sizes: {
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      large: '1.8rem',
      xlarge: '2.0rem',
      xxlarge: '2.8rem',
      xxxlarge: '3.6rem',
      huge: '5.2rem'
    }
  },
  colors: {
    primary: '#E65A6D',
    primaryColor: {
      primary100: '#E65A6D1A',
      primary200: '#E65A6D33',
      primary300: '#E65A6D4D',
      primary400: '#E65A6D66',
      primary500: '#E65A6D80',
      primary600: '#E65A6D99',
      primary700: '#E65A6DB3',
      primary800: '#E65A6DCC',
      primary900: '#E65A6DE6',
      primary1000: '#E65A6D'
    },
    secondary: '#C9FCC1',
    secondaryColor: {
      secondary100: '#C9FCC11A',
      secondary200: '#C9FCC133',
      secondary300: '#C9FCC14D',
      secondary400: '#C9FCC166',
      secondary500: '#C9FCC180',
      secondary600: '#C9FCC199',
      secondary700: '#C9FCC1B3',
      secondary800: '#C9FCC1CC',
      secondary900: '#C9FCC1E6',
      secondary1000: '#C9FCC1'
    },
    mainBg: '#030517',
    lightBg: '#34C1FF66',
    matisse: '#1F688E',
    white: '#FAFAFA',
    black: '#030517',
    neutral1: '#2C3A4B',
    neutral: {
      white: '#FFFFFF',
      neutral100: '#F4F6F9',
      neutral200: '#EBEEF2',
      neutral300: '#DADEE3',
      neutral400: '#A5ABB3',
      neutral500: '#858C94',
      neutral600: '#6D7580',
      neutral700: '#545D69',
      neutral800: '#394452',
      neutral900: '#2C3A4B',
      black: '#09101D'
    },
    lightGray: '#EAEAEA',
    gray: '#8F8F8F',
    darkGray: '#2E2F42',
    red: '#FF6347',
    success: '#287D3C',
    error: '#DA1414',
    warning: '#B95000'
  },
  accentColors: {
    blue: '#2E8EF7',
    green: '#44cc44',
    orange: '#FFC371',
    pink: '#FF409A',
    purple: '#C341FF'
  },
  gradients: {
    lighterBlue: `linear-gradient(180deg, rgba(54, 200, 255, 0.06) 15.1%, rgba(196, 196, 196, 0) 63.54%)`,
    blue: `linear-gradient(222deg, #36CBFF -1.24%, #2E8EF7 111.42%);`,
    lightBlue: `linear-gradient(109deg, #FFFFFF 0%, #A3DEFD 101.14%);`,
    green: `linear-gradient(213deg, #16B4EB 4.85%, #23E73C 119.17%);`,
    orange: `linear-gradient(135deg, #FF5F6D 2%, #FFC371 100%);`,
    pink: `linear-gradient(136deg, #FF409A 8.34%, #C438EF 95.26%);`,
    purple: `linear-gradient(213.66deg, #8E31FD 4.85%, #C341FF 119.17%);`
  },
  spacings: {
    xxsmall: '0.8rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  },
  shadow: {
    fluffy: {
      elevation5: '0 15px 60px #0000001f;'
    }
  }
} as const
