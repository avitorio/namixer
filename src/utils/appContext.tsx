import {
  Dispatch,
  SetStateAction,
  useState,
  useContext,
  createContext,
  ReactNode
} from 'react'

export interface IProviderProps {
  children?: ReactNode
}

type AppContextState = { domain: string }

type AppContextValue = {
  state: AppContextState
  setState: Dispatch<SetStateAction<AppContextState>>
}

export const AppContext = createContext<AppContextValue | undefined>(undefined)

export const AppProvider = (props: IProviderProps) => {
  const [state, setState] = useState({ domain: 'YourDomain.com' })

  return (
    <AppContext.Provider value={{ state, setState }}>
      {props.children}
    </AppContext.Provider>
  )
}

export function useAppContext() {
  const ctxValue = useContext(AppContext)
  if (ctxValue === undefined)
    throw new Error(
      'Expected an AppProvider somewhere in the react tree to set context value'
    )
  return ctxValue // now has type AppContextValue
  // or even provide domain methods for better encapsulation
}
