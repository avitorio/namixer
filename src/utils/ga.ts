// https://developers.google.com/analytics/devguides/collection/gtagjs/pages
export const pageview = (url: URL): void => {
  window.gtag('config', `${process.env.NEXT_PUBLIC_GA_TRACKING}`, {
    page_path: url
  })
}

type EventParams = { [key: string]: string | number }

type GTagEvent = {
  event: string
  params: EventParams
}

// https://developers.google.com/analytics/devguides/collection/gtagjs/events
export const event = ({ event, params }: GTagEvent): void => {
  window.gtag('event', event, {
    ...params
  })
}
