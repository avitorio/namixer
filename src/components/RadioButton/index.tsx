import { InputHTMLAttributes } from 'react'

import * as S from './styles'

type RadioButtonValue = string | ReadonlyArray<string> | number

export type RadioButtonProps = {
  onCheck?: (value?: RadioButtonValue) => void
  label?: string
  labelColor?: 'white' | 'black'
  labelFor?: string
  value?: RadioButtonValue
} & InputHTMLAttributes<HTMLInputElement>

const RadioButton = ({
  label,
  onCheck,
  labelFor = '',
  value,
  ...props
}: RadioButtonProps) => {
  const onChange = () => {
    !!onCheck && onCheck(value)
  }

  return (
    <S.Wrapper>
      <S.Input
        id={labelFor}
        type="radio"
        value={value}
        onChange={onChange}
        {...props}
      />
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
    </S.Wrapper>
  )
}

export default RadioButton
