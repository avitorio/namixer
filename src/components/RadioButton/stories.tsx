import { Story, Meta } from '@storybook/react/types-6-0'
import RadioButton, { RadioButtonProps } from '.'

export default {
  title: 'Form/RadioButton',
  component: RadioButton,
  parameters: {
    layout: 'fullscreen',
    backgrounds: {
      default: 'won-dark'
    }
  },
  argTypes: {
    onCheck: { action: 'checked' }
  }
} as Meta

export const Default: Story<RadioButtonProps> = (args) => (
  <>
    <div style={{ padding: 10 }}>
      <RadioButton
        label="primeiro"
        labelFor="primeiro"
        id="primeiro"
        name="nome"
        value="primeiro"
        defaultChecked
        {...args}
      />
    </div>
    <div style={{ padding: 10 }}>
      <RadioButton
        label="segundo"
        labelFor="segundo"
        id="segundo"
        name="nome"
        value="segundo"
        {...args}
      />
    </div>
    <div style={{ padding: 10 }}>
      <RadioButton
        label="terceiro"
        labelFor="terceiro"
        id="terceiro"
        name="nome"
        value="terceiro"
        {...args}
      />
    </div>
  </>
)
