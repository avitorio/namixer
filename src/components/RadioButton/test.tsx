import { render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'
import theme from 'styles/theme'

import RadioButton from '.'

describe('<RadioButton />', () => {
  it('should render with label (white)', () => {
    const { container } = render(
      <RadioButton label="RadioButton" labelFor="check" value="anyValue" />
    )

    const label = screen.getByText('RadioButton')
    expect(label).toBeInTheDocument()
    expect(label).toHaveStyle({ color: theme.colors.neutral.neutral900 })
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render without label', () => {
    render(<RadioButton />)

    expect(screen.queryByLabelText('RadioButton')).not.toBeInTheDocument()
  })

  it('should dispatch onCheck when label status changes', async () => {
    const onCheck = jest.fn()
    render(
      <RadioButton
        label="RadioButton"
        labelFor="RadioButton"
        onCheck={onCheck}
        value="anyValue"
      />
    )

    expect(onCheck).not.toHaveBeenCalled()

    userEvent.click(screen.getByLabelText('RadioButton'))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith('anyValue')
  })

  it('Should be accessible with tab', () => {
    render(<RadioButton label="RadioButton" labelFor="RadioButton" />)

    const radioButton = screen.getByLabelText('RadioButton')

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(radioButton).toHaveFocus()
  })
})
