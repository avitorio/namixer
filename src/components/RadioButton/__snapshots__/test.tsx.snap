// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton /> should render with label (white) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  opacity: 0;
  position: fixed;
  width: 0px;
}

.c1:checked + label {
  border: 2px solid #44cc44;
  background: #C9FCC1;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #F4F6F9;
  border: 2px solid #ffffff00;
  border-radius: 0.4rem;
  color: #2C3A4B;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Source Sans Pro,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-weight: 600;
  padding: 1.6rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c2:hover {
  background: #f3fef2;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="check"
    type="radio"
    value="anyValue"
  />
  <label
    class="c2"
    for="check"
  >
    RadioButton
  </label>
</div>
`
